### PR TITLE
API and handler for creating job from definition

### LINKED ISSUE #209: Run now from a job definition
### Problem

Previously, we offered a way for users to "re-run" a previously run job, but this didn't leverage a job definition. We removed that feature as it was incompatible with the API change we were making (removing the `input_url`). However, we believe that users we will still want to rerun a job. 

### Proposed Solution

We proposed allowing users to "re-run" a job by adding a "Run now" action for a job definition. To implement this we would create a new REST API and `Scheduler` method and then build the UI on top of this. The reason we are thinking of using the job definition for this is as follows:

* The job definition already has an immutable snapshot of the input file that can be used for the job.
* The job definition allows us to show the user all of the jobs that are "grouped" by the job definition.

### PR DESCRIPTION
**Pull request by 3coins**
Fixes #209 

Added a new API and handler to create job from job definition. 

#### Scheduler API
```py
def create_job_from_definition(self, job_definition_id: str, model: CreateJobFromDefinition):
        """Creates a new job based on a job definition"""
        raise NotImplementedError("must be implemented by subclass")
```

#### Handler
```
POST /scheduler/job_definitions/<job_definition_id>/jobs
```

### CODE REVIEW

File: jupyter_scheduler/handlers.py [hunk 1]
```diff
@@ -206,6 +207,33 @@ async def delete(self, job_id):
...
+class JobFromDefinitionHandler(ExtensionHandlerMixin, JobHandlersMixin, APIHandler):
+    @tornado.web.authenticated
+    async def post(self, job_definition_id: str):
+        payload = self.get_json_body()
+        try:
+            model = CreateJobFromDefinition(**payload)
+            job_id = await ensure_async(
+                self.scheduler.create_job_from_definition(job_definition_id, model=model)
+            )
+        except InputUriError as e:
```
Comment (ellisonbg) [inline jupyter_scheduler/handlers.py:219]: It doesn't hurt this to be there, but haven't we removed the input URI altgoether?

File: jupyter_scheduler/models.py [hunk 2]
```diff
@@ -269,6 +269,10 @@ class ListJobDefinitionsResponse(BaseModel):
     next_token: Optional[str] = None
 
 
+class CreateJobFromDefinition(BaseModel):
```
Comment (ellisonbg) [inline jupyter_scheduler/models.py:272]: I think this is a good start to the things that users will want to customize for this usage case.

File: jupyter_scheduler/scheduler.py [hunk 3]
```diff
@@ -608,6 +613,17 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
 
         return list_response
 
+    def create_job_from_definition(self, job_definition_id: str, model: CreateJobFromDefinition):
+        job_id = None
+        definition = self.get_job_definition(job_definition_id)
+        if definition and definition.active:
```
Comment (ellisonbg) [inline jupyter_scheduler/scheduler.py:619]: My thought is that the run now from a job def should work even if the job def is paused. Thoughts?

File: jupyter_scheduler/scheduler.py [hunk 4]
```diff
@@ -608,6 +613,17 @@ def list_job_definitions(self, query: ListJobDefinitionsQuery) -> ListJobDefinit
...
+            input_uri = self.get_staging_paths(definition)["input"]
```
Comment (ellisonbg) [inline jupyter_scheduler/scheduler.py:620]: Does this end up copying the input file to the staging directory for the job?

File: jupyter_scheduler/task_runner.py [hunk 5]
```diff
@@ -260,15 +260,7 @@ def delete_job_definition(self, job_definition_id: str):
...
-        definition = self.scheduler.get_job_definition(job_definition_id)
-        if definition and definition.active:
-            input_uri = self.scheduler.get_staging_paths(definition)["input"]
-            self.scheduler.create_job(
-                CreateJob(
-                    **definition.dict(exclude={"schedule", "timezone"}, exclude_none=True),
-                    input_uri=input_uri
-                )
-            )
+        self.scheduler.create_job_from_definition(job_definition_id, CreateJobFromDefinition())
```
Comment (ellisonbg) [inline jupyter_scheduler/task_runner.py:263]: If we remove the is active test from `create_job_from_definition` do we need it here?

**Review by ellisonbg — COMMENTED** [review]
Overall looks good. I left some minor comments/questions.

File: jupyter_scheduler/scheduler.py (hunk 3 above)
Comment (andrii-i) [reply jupyter_scheduler/scheduler.py:619]: I think run now should work even if the job def is paused as I can see scenarios where users would want to run a job off paused definition without restarting it (e.g. if it's a resource/time-heavy job with a high-frequency schedule)

File: jupyter_scheduler/handlers.py (hunk 1 above)
Comment (3coins) [reply jupyter_scheduler/handlers.py:219]: Correct, this should not apply here.

File: jupyter_scheduler/scheduler.py (hunk 3 above)
Comment (3coins) [reply jupyter_scheduler/scheduler.py:619]: Agree, we are reusing this from task runner as well, so this was a copy. Will correct this.

File: jupyter_scheduler/scheduler.py (hunk 4 above)
Comment (3coins) [reply jupyter_scheduler/scheduler.py:620]: Yes, this references the create_job method which copies the file.

File: jupyter_scheduler/task_runner.py (hunk 5 above)
Comment (3coins) [reply jupyter_scheduler/task_runner.py:263]: Correct, this needs an update.

File: jupyter_scheduler/scheduler.py (hunk 3 above)
Comment (3coins) [reply jupyter_scheduler/scheduler.py]: Updated.

File: jupyter_scheduler/task_runner.py (hunk 5 above)
Comment (3coins) [reply jupyter_scheduler/task_runner.py]: Updated

File: jupyter_scheduler/handlers.py [hunk 6]
```diff
@@ -206,6 +207,27 @@ async def delete(self, job_id):
...
+            job_id = await ensure_async(
+                self.scheduler.create_job_from_definition(job_definition_id, model=model)
+            )
+        except SchedulerError as e:
+            self.log.exception(e)
+            raise tornado.web.HTTPError(500, str(e)) from e
+        except Exception as e:
+            self.log.exception(e)
+            raise tornado.web.HTTPError(
+                500, "Unexpected error occurred during creation of Job."
```
Comment (jweill-aws) [inline jupyter_scheduler/handlers.py:225]: Minor: Here and elsewhere, `job` should be in lowercase.

**Review by jweill-aws — APPROVED** [review]

**Pull request merged**